ivadomed/inference.py

```diff
@@ -86,7 +86,23 @@ def pred_to_nib(data_lst, z_lst, fname_ref, fname_out, slice_axis, debug=False,
                                               nib_ref.header['pixdim'][1:4],
                                               fname_prefix)
         arr_pred_ref_space = postpro.apply()
-    nib_pred = nib.Nifti1Image(arr_pred_ref_space, nib_ref.affine)
+    # TODO: remove the "[0:3]" in the test if test below. The reason I needed to add it is because in some cases, the
+    #  prediction has the shape (nx, ny, nz, 1) while the inpute image is (nx, ny, nz), even for a one-channel
+    #  prediction. Consequently, the if check below goes to the "else" case, and
+    #  issue https://github.com/ivadomed/ivadomed/issues/711 is still happening.
+    #  I'm afraid this [0:3] hack will break for 2D data. Also, we might want to make sure the output arr_pred_ref_space
+    #  always has the same shape as the input image. For multiple class prediction, maybe we can find another way
+    #  to store the predictions (e.g. list of np arrays?).
+    if arr_pred_ref_space.shape[0:3] == nib_ref.get_fdata().shape[0:3]:
```

Comment:
From what I understand in the description above, if the dimensions of the prediction is (nx, ny, nz, 1) while the input image is (nx, ny, nz) the code should go to the "else" statement (correct me if I'm wrong). It seems to me that in all cases the code under `if arr_pred_ref_space.shape[0:3] == nib_ref.get_fdata().shape[0:3]:` will be ran since `[nx, ny, nz, 1][0:3] == [nx, ny, nz][[0:3] `. Maybe this bit of code is in the case of a prediction with the dimensions (1, nx, ny, nz)? 

From this error https://github.com/ivadomed/ivadomed/pull/714/checks?check_run_id=2107966525, the problem seems to come from the `save_color_labels` function which generates nifti files with non-int or float data. Rather than introducing this condition that might cause errors with 2D images, I'd simply comment the lines related to this function (which simply creates an extra file that we barely use) and adjust the function `save_color_labels`  in a future PR. If this plans work for you, I can take care of it.

Reply:
> if the dimensions of the prediction is (nx, ny, nz, 1) while the input image is (nx, ny, nz) the code should go to the "else" statement (correct me if I'm wrong). 

sorry i wasn't clear-- let me know if this is clearer-- if it is i will update the code with this comment below:
~~~
    # TODO: remove the "[0:3]" in "if test" below. The reason I added it is because in some cases, the
    #  prediction has the shape (nx, ny, nz, 1) while the input image is (nx, ny, nz), even for a one-channel
    #  prediction. Consequently, the "if check" below goes to the "else" case, and
    #  issue https://github.com/ivadomed/ivadomed/issues/711 is still happening.
    #  What we want is: if the output prediction has the same (nx, ny, nz) as the reference image, the we need
    #  to copy the header (not just the affine), in order to fix #711.
    #  However, I am afraid this "[0:3] hack" will break for 2D data. Also, we might want to make sure the output arr_pred_ref_space
    #  always has the same shape as the input image. For multiple class prediction, maybe we can find another way
    #  to store the predictions (e.g. list of np arrays?).
~~~

Reply:
> From this error https://github.com/ivadomed/ivadomed/pull/714/checks?check_run_id=2107966525, the problem seems to come from the save_color_labels function which generates nifti files with non-int or float data. Rather than introducing this condition that might cause errors with 2D images, I'd simply comment the lines related to this function (which simply creates an extra file that we barely use) and adjust the function save_color_labels in a future PR. If this plans work for you, I can take care of it.

that would be awesome! yes-- i am completely swamped with other tasks now

Reply:
so IIUC: what you are suggesting is that after the fix you suggest, `arr_pred_ref_space.shape` will ALWAYS be the same as `nib_ref.get_fdata().shape`?  glancing at the code, my understanding was that in some cases the 4th dimension is used for different channels (so you suggest to not do that anymore?)

Reply:
Yes, exactly! The prediction and the original image will always have different dimensions when using ivadomed as we store the different classes in the 4th dimension. Changing this would require an important refactoring. However, in SCT we always store the data in different files (one file per class) so this shape discrepancy should never happen when calling segment_volume from SCT. I'll do some test to verify if this is true. If so, I think only this line is required `nib_pred = nib.Nifti1Image(arr_pred_ref_space, None, nib_ref.header.copy())` since this code still works when the shapes are different (even though it might generate the qform and form codes, but with the current form of the code I think it is inevitable).

Reply:
> since this code still works when the shapes are different

when i tested it, the following code:
~~~
nib_pred = nib.Nifti1Image(arr_pred_ref_space, None, nib_ref.header.copy())
~~~

will produce an error if e.g. one is (nx, ny, nz) and the other is (nx, ny, nz, 1). Hence my "hack"

Reply:
If there is a concern that this "[0:3] hack" will break for 2D data, should we add a test for 2D data?

Reply:
I was wrong, the predictions generated from SCT initially have a shape of (nx, ny, nz, X) and are converted to a list afterward: https://github.com/ivadomed/ivadomed/blob/master/ivadomed/inference.py#L300. There will always be a shape discrepancy in the `pred_to_nib` function. 

I tested the code in its actual form (with the header copied even if the shapes are different), and I got the following qform/sform codes: 
![image](https://user-images.githubusercontent.com/49137243/111193368-79ca0680-8590-11eb-9913-c276bb5ca0f8.png)

The expected behavior is that qform and sform should always match the original image?

Reply:
> The expected behavior is that qform and sform should always match the original image?

that is correct.

Reply:
The current version of the code doesn't fix the issue regarding the qform and sform codes I believe. See suggestion below.